### PR TITLE
fix: keep the same instance of content when revalidating

### PR DESCRIPTION
**Pull request by gdorsi**
Since now we had this weird thing that, when the group updates a new "currentContent" is created.

This has always been a pain while testing, because something like this would fail in an hard-to-understand way:
```ts
test("the map should rebuild when the group permissions change", async () => {
  const alice = setupTestNode({
    connected: true,
  });
  const bob = setupTestNode({
    connected: true,
  });
  const group = alice.node.createGroup();
  const map = group.createMap();
  
  group.addMember("everyone", "writer");  // new content here!

  const mapOnBob = await loadCoValueOrFail(bob.node, map.id);
  map.set("fromBob", true);

  await waitFor(() => {
    expect(map.get("fromBob")).toBeTrue() // This is no more the current content, so it won't be never updated
  });
});
```

The same thing happens on `jazz-tools`.

In this PR I'm changing the approach to reset the state of the current view, keeping the same reference.
Had to introduce a `version` field to let SubscriptionScope know when we've got a "revalidation" because only checking the `totalValidTransactions` is not enough for some edge cases (when after the revalidation we have the same amount of valid transactions but different content, yes we already have a test covering this case).

We've got no one complaining about this, besides me, but since it can make things fail in an unpredictable way I think it's time to improve the flow.

### CODE REVIEW

File: packages/cojson/src/coValues/coStream.ts [hunk 1]
```diff
@@ -69,6 +69,14 @@ export class RawCoStreamView<
     this.processNewTransactions();
   }
 
+  rebuildFromCore() {
+    this.version++;
+    this.items = {};
+    this.knownTransactions = { [this.core.id]: 0 };
+    this.totalValidTransactions = 0;
+    this.processNewTransactions();
```
Comment (nrainhart) [inline packages/cojson/src/coValues/coStream.ts]: This is the same thing we do in the constructor, right? Since we can't extract it into a function, we should add a comment to avoid forgetting to keep both in sync

File: packages/cojson/src/coValues/coMap.ts [hunk 2]
```diff
@@ -353,18 +354,7 @@ export class RawCoMapView<
       listener(core.getCurrentContent() as this);
     });
   }
-}
 
-/** A collaborative map with precise shape `Shape` and optional static metadata `Meta` */
-export class RawCoMap<
```
Comment (nrainhart) [inline packages/cojson/src/coValues/coMap.ts:359]: No more split between read & write methods? I don't have a strong opinion either way, but curious if there was any reason to do this now

File: packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts [hunk 3]
```diff
@@ -172,6 +173,7 @@ export class SubscriptionScope<D extends CoValue> {
     } else {
       const hasChanged =
         update.totalValidTransactions !== this.totalValidTransactions ||
+        update.version !== this.version ||
         // Checking the identity of the raw value makes us cover the cases where the group
         // has been updated and the coValues that don't update the totalValidTransactions value (e.g. FileStream)
         this.value.value.$jazz.raw !== update;
```
Comment (nrainhart) [inline packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts]: Is this still necessary?

File: packages/cojson/src/tests/sync.load.test.ts [hunk 4]
```diff
@@ -481,10 +481,7 @@ describe("loading coValues from server", () => {
     map.set("fromServer", "updated", "trusting");
 
     await waitFor(() => {
-      const coValue = expectMap(
-        client.node.expectCoValueLoaded(map.id).getCurrentContent(),
-      );
-      expect(coValue.get("fromServer")).toEqual("updated");
+      expect(map.get("fromServer")).toEqual("updated");
```
Comment (nrainhart) [inline packages/cojson/src/tests/sync.load.test.ts:484]: ✨

**Review by nrainhart — APPROVED** [review]
Nice!!

File: packages/cojson/src/coValues/coMap.ts (hunk 2 above)
Comment (gdorsi) [reply packages/cojson/src/coValues/coMap.ts:359]: Because this was making the refactoring to make `processNewTransactions` and `rebuildFromCore` more complicated.

I don't have a strong opinion either, but I've never got value from this separation and it's only been an hindrance so far.

Good observation thou, I should have added context on the PR for this change.

File: packages/cojson/src/coValues/coStream.ts (hunk 1 above)
Comment (gdorsi) [reply packages/cojson/src/coValues/coStream.ts]: Good point, going to try to extract this into a function.

File: packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts (hunk 3 above)
Comment (gdorsi) [reply packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts]: Probably no, let's try to remove it!

File: packages/cojson/src/coValues/coStream.ts (hunk 1 above)
Comment (nrainhart) [reply packages/cojson/src/coValues/coStream.ts]: I don't think you'll be able to. Typescript is very strict with field assignments in constructors 👎

Comment (gdorsi) [reply packages/cojson/src/coValues/coStream.ts]: This should make it more clear that internal state needs to be part of the reset: https://github.com/garden-co/jazz/pull/3162/commits/94afaebf462eee9cbbe23012e5a5ba503c67cf05

File: packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts (hunk 3 above)
Comment (gdorsi) [reply packages/jazz-tools/src/tools/subscribe/SubscriptionScope.ts]: Removed: https://github.com/garden-co/jazz/pull/3162/commits/c3cd9127ebfa8fdc8b8956a3623996c27b742647
